### ISSUE TITLE
Add content-defined chunking for better deduplication

### ISSUE THREAD

**Issue by Harliff** [openzfs/zfs#11400]
It would be great to be able to use CDC (content-defined chunking) for deduplication instead of FSC (fix-sized chunking) to get better dedupratio.

Links:
https://en.wikipedia.org/wiki/Rolling_hash
https://www.sciencedirect.com/science/article/abs/pii/S0167739X18320053

**Comment by RubenKelevra** [comment]
Agreed. IPFS uses Buzhash which replaced the Rabin chunker they used before, which was somewhat slow.

It works great to deduplicate data that has slightly shifted, without knowing their alignment. Like updated binaries or VM images of different virtual machines.

I'm currently using IPFS on ZFS on multiple machines. I'm holding some data inside and outside of IPFS, which means I have to stick to 256 KByte chunking on ZFS and IPFS to have ZFS properly deduplicate those data.

When ZFS would support Buzhash I could use Buzhash in IPFS as well :)

This is how Buzhash is configured (and implemented) in IPFS:

https://github.com/ipfs/go-ipfs-chunker/blob/master/buzhash.go

**Comment by Harliff** [comment]
Just found info about 'casync' synchronization tool (http://0pointer.net/blog/casync-a-tool-for-distributing-file-system-images.html) and 'FastCDC' algorithm (https://www.usenix.org/system/files/conference/atc16/atc16-paper-xia.pdf) which is used in the casync.

Maybe the 'casync' approaches to data deduplication can be useful when implementing the CDC in the ZFS.

**Comment by nrdxp** [comment]
> Maybe the 'casync' approaches to data deduplication can be useful when implementing the CDC in the ZFS.

Incidentally, I was recently thinking how nice it would be if a filesystem would dedup in a way that is compatible with casync/[desync](https://github.com/folbricht/desync) so that you could just send and receive chunks over the network without any extra work or space.
